packages/astro/src/core/render/dev/index.ts

```diff
@@ -44,36 +41,31 @@ export interface SSROptions {
 	middleware?: AstroMiddlewareInstance<unknown>;
 }
 
-export type ComponentPreload = [SSRLoadedRenderer[], ComponentInstance];
-
-export async function loadRenderers(
-	moduleLoader: ModuleLoader,
-	settings: AstroSettings
-): Promise<SSRLoadedRenderer[]> {
-	const loader = (entry: string) => moduleLoader.import(entry);
-	const renderers = await Promise.all(settings.renderers.map((r) => loadRenderer(r, loader)));
-	return filterFoundRenderers(renderers);
-}
-
 export async function preload({
 	env,
 	filePath,
-}: Pick<SSROptions, 'env' | 'filePath'>): Promise<ComponentPreload> {
+}: {
+	env: DevelopmentEnvironment;
+	filePath: URL;
+}): Promise<ComponentInstance> {
 	// Important: This needs to happen first, in case a renderer provides polyfills.
-	const renderers = await loadRenderers(env.loader, env.settings);
+	const renderers = await loadRenderers(env.settings, env.loader);
+	// Override the environment's renderers. This ensures that if renderers change (HMR)
+	// The new instances are passed through.
+	env.renderers = renderers;
 
 	try {
 		// Load the module from the Vite SSR Runtime.
 		const mod = (await env.loader.import(viteID(filePath))) as ComponentInstance;
 
-		return [renderers, mod];
+		return mod;
```

Comment:
With this simplification, we can directly return the `mod` instead. The rest of the changes are to accommodate this type change.